cirq/circuits/circuit_operation.py

```diff
@@ -0,0 +1,382 @@
+# Copyright 2020 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A structure for encapsulating entire circuits in an operation.
+
+A CircuitOperation is an Operation object that wraps a FrozenCircuit. When
+applied as part of a larger circuit, a CircuitOperation will execute all
+component operations in order, including any nested CircuitOperations.
+"""
+
+from typing import TYPE_CHECKING, AbstractSet, Dict, List, Tuple, Union
+
+import numpy as np
+
+from cirq import linalg, ops, protocols, study
+from cirq._compat import proper_repr
+
+if TYPE_CHECKING:
+    import cirq
+
+
+INT_TYPE = Union[int, np.integer]
+
+
+class CircuitOperation(ops.Operation):
+    """An operation that encapsulates a circuit.
+
+    This class captures modifications to the contained circuit, such as tags
+    and loops, to support more condensed serialization. Similar to
+    GateOperation, this type is immutable.
+    """
+
+    def __init__(self, circuit: 'cirq.FrozenCircuit') -> None:
```

Comment:
My intent here was to emphasize reusability of the `FrozenCircuit` since that's a key feature of `CircuitOperation`, but I can see how accepting `OP_TREE` would be more user-friendly. I'll look at this alongside @maffoo's recommendations for the constructor.

Reply:
Sounds good. If I had to choose between the comments on the PR, this would be one to add.

Reply:
This is now supported along with the original `circuit` parameter.